### PR TITLE
Support stack traces with gdb (and debugger invocation)

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: LIB_MODE=shared is much more efficient for building all the unit tests but comes with the downside of ugly stack traces, generally missing name demangling and source line info. Searching the internet suggests the reliable way to get stack traces with dynamic loading is with gdb.

This change automatically tries to use gdb to get a stack trace if built with LIB_MODE=shared, and only on Linux because that's where we have the capability to attach to the proper thread. (We could revise the exact conditions in the future.) If there's a failure invoking gdb, it falls back on the old method. Obscure details of making the output reasonable / pretty are in the source code comments.

Based on this, it was easy to make it so that running a test command with ROCKSDB_DEBUG=1 would invoke gdb whenever the stack trace handler was invoked, so I included that.

Intended follow-up: make LIB_MODE=shared the new default `make` build config

Test Plan: manual, mostly by injecting an "assert(false)" into a unit test and trying different build modes etc.

Although gdb is slower to start showing stack trace output, it seems overall faster in many if not most cases, presumably because it doesn't reload the symbol table for each stack entry. At least with parallel test runs, having many tests dumping stacks with the old method can take so long it appears to hang the test run.

### CODE REVIEW

File: port/stack_trace.cc [hunk 1]
```diff
@@ -135,6 +137,68 @@ void PrintStack(void* frames[], int num_frames) {
...
+  bool debug = debug_env != nullptr && strlen(debug_env) > 0;
+
+  if (dll || debug) {
+    // Allow ouside debugger to attach, even with Yama security restrictions
+#ifdef PR_SET_PTRACER_ANY
+    (void)prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY, 0, 0, 0);
+#endif
+    // Try to invoke GDB, either for stack trace or debugging
+    char pid_str[20];
+    snprintf(pid_str, sizeof(pid_str), "%lld", (long long)getpid());
```
Comment (pdillinger) [inline port/stack_trace.cc:159]: Btw, I don't think heap allocation is supported in signal handlers, which is why we use stack-allocated buffers and mostly plain C.

File: port/stack_trace.cc [hunk 2]
```diff
@@ -135,6 +137,68 @@ void PrintStack(void* frames[], int num_frames) {
...
+      // child process
+      if (debug) {
+        fprintf(stderr, "Invoking GDB for debugging (ROCKSDB_DEBUG=%s)...\n",
+                debug_env);
+        execlp(/*cmd in PATH*/ "gdb", /*arg0*/ "gdb", "-p", pid_str,
+               (char*)nullptr);
+        return;
+      } else {
+        fprintf(stderr, "Invoking GDB for stack trace...\n");
+        // TODO: does this always select the correct thread in gdb?
```
Comment (cbi42) [inline port/stack_trace.cc:171]: From my limited manual testing, it seems that main thread's backtrace is always printed even when assertion failure is in compaction threads. Researching on the internet does not give a solution to attach to the failing thread automatically. I wonder if it still makes sense to change default to shared library for CI tests as the backtrace may give limited information (the old method prints backtrace for the failing thread).

**Review by cbi42 — APPROVED** [review]
LGTM! I can see `ROCKSDB_DEBUG` being very helpful. The comment is more for the other PR that makes LIB_MODE=shared the default.

File: port/stack_trace.cc [hunk 3]
```diff
@@ -135,6 +137,68 @@ void PrintStack(void* frames[], int num_frames) {
 }
 
 void PrintStack(int first_frames_to_skip) {
+#ifdef ROCKSDB_DLL
+  // LIB_MODE=shared build produces modiocre information from the above
```
Comment (laurynas-biveinis) [inline port/stack_trace.cc:141]: "modiocre"

**Comment by pdillinger** [comment]
It works 😁 https://app.circleci.com/pipelines/github/facebook/rocksdb/24102/workflows/e49e2560-89ab-4972-ad4f-acb5e5a3e109/jobs/565036

**Comment by cbi42** [comment]
> It works 😁 https://app.circleci.com/pipelines/github/facebook/rocksdb/24102/workflows/e49e2560-89ab-4972-ad4f-acb5e5a3e109/jobs/565036

Nice!

**Comment by cbi42** [comment]
Any idea on why stack trace / gdb attaching does not work with db_stress? I made added the following assertion:
```
--- a/db/db_impl/db_impl_write.cc
+++ b/db/db_impl/db_impl_write.cc
@@ -307,7 +307,7 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
                         disable_memtable, batch_cnt, pre_release_callback,
                         post_memtable_callback);
   StopWatch write_sw(immutable_db_options_.clock, stats_, DB_WRITE);
-
+  assert(false);
   write_thread_.JoinBatchGroup(&w);
   if (w.state == WriteThread::STATE_PARALLEL_MEMTABLE_WRITER) {
     // we are a non-leader in a parallel group
```

and got the following output with or without ROCKSDB_DEBUG=1 

```
2023/02/03-11:27:22  Initializing worker threads
2023/02/03-11:27:22  Starting database operations
db_stress: db/db_impl/db_impl_write.cc:310: rocksdb::Status rocksdb::DBImpl::WriteImpl(const rocksdb::WriteOptions&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, uint64_t*, uint64_t, bool, uint64_t*, size_t, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*): Assertion `false' failed.
db_stress: db/db_impl/db_impl_write.cc:310: rocksdb::Status rocksdb::DBImpl::WriteImpl(const rocksdb::WriteOptions&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, uint64_t*, uint64_t, bool, uint64_t*, size_t, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*): Assertion `false' failed.
db_stress: db/db_impl/db_impl_write.cc:310: rocksdb::Status rocksdb::DBImpl::WriteImpl(const rocksdb::WriteOptions&, rocksdb::WriteBatch*, rocksdb::WriteCallback*, uint64_t*, uint64_t, bool, uint64_t*, size_t, rocksdb::PreReleaseCallback*, rocksdb::PostMemTableCallback*): Assertion `false' failed.
Received signal 6 (Aborted)
Aborted (core dumped)
```

**Comment by pdillinger** [comment]
> Any idea on why stack trace / gdb attaching does not work with db_stress? I added the following assertion:

The problem here is that many threads are hitting the failure and I think that causes signal after first one enters the signal handler, which leads to abrupt abort. If you put the `assert(false)` after `assert(w.state == WriteThread::STATE_GROUP_LEADER);` to ensure only one thread is affected, it works. I looked at signal handling APIs but didn't see a solution to this problem.

**Pull request closed without merging**